### PR TITLE
feat(host): add host certificate collector and analyzer

### LINKED ISSUE #762: Collector & Analyzer for certificates
**Describe the rationale for the suggested feature.**

Several support cases have come through where certificates have expired, we would like to be able to explain this at the analyzer stage so folks know immediately what to do about it.

**Describe the feature**

Ensure that the certificate expiry information is available in support bundles for:
* Kubernetes
* weave
* kURL registry
* Envoy

Create an analyzer that will highlight expired certificates from any of the above, along with a URL that will show instructions for remedy.

**Describe alternatives you've considered**

.

**Additional context**

Related Shortcut: https://app.shortcut.com/replicated/story/42738/add-collectors-and-analyzers-for-expired-certificates

Envoy: https://community.replicated.com/t/kurl-how-to-manually-rotate-an-expired-certificate-for-envoy/889

kURL Registry: https://community.replicated.com/t/what-to-do-if-the-kurl-registry-certificates-have-expired/955

Contour: https://community.replicated.com/t/kots-is-it-possible-to-configure-tls-certs-for-contour-in-the-embedded-installation/390/2

Checking the Kubernetes certs themselves is more difficult as it requires `sudo kubeadm certs check-expiration` which requires sudo.

**Definition of done**

- [x] Write a spec for the collectors and analysers - Done [here](https://github.com/replicatedhq/troubleshoot/issues/762#issuecomment-1285916327)
- [x] Develop an in-cluster collector
  - [x] Certs from secrets
  - [x] Certs from config maps
  - [ ] Certs from endpoints
  - [x] In-cluster kubernetes cluster certs - most are in `kube-*` config maps
- [x] Develop a host collector
  - [x] Kubernetes cluster certs from host paths i.e `/etc/kubernetes/pki`, `/var/lib/kubelet/pki` on a kubeadm-based cluster
  - [ ] Certs from endpoints
  - [x] Certs from a list of paths in the host e.g `/etc/myservice/*.crt`
- [x] Develop an analyser - reviews expiry amongst other things
- [ ] The default specs we provide in [troubleshoot-specs](https://github.com/replicatedhq/troubleshoot-specs) including analyzers for the above certificates with links to docs to fix them. Collect the following
  - [ ] Kubernetes (host & in-cluster collectors)
  - [ ] kURL registry
  - [ ] Contour/envoy
  - [ ]  KOTS Admin Console (kotsadm-tls)
- [x] Document both the collector and analyser in [troubleshoot.sh docs](https://troubleshoot.sh/docs/)

### PR DESCRIPTION
**Pull request by DexterYan**
## Description, Motivation and Context

Please include a summary of the change or what problem it solves. Please also include relevant motivation and context.

issue: #762

# Support Bundle Yaml
```
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: certificate
spec:
  hostCollectors:
    - certificatesCollection:
        paths: 
        - ~/apiserver-kubelet-client.crt
        - /etc/ssl/corp.crt
        - ~/ca.crt
  hostAnalyzers:
    - certificatesCollection:
        outcomes:
          - pass:
              message: Certificate is valid
          - warn:
              when: "notAfter < Today + 365 days"
              message: Certificate is about to expire
          - fail:
              when: "notAfter < Today"
              message: Certificate is expired
```

# Result Json
```
[
    {
        "name": "host.cerfiticates.collection",
        "labels": {
            "desiredPosition": "1",
            "iconKey": "",
            "iconUri": ""
        },
        "insight": {
            "name": "host.cerfiticates.collection",
            "labels": {
                "iconKey": "",
                "iconUri": ""
            },
            "primary": "Host Cerfiticates Collection",
            "detail": "Certificate is valid, obtained from /Users/ubuntu/dev/replicated/troubleshoot/apiserver-kubelet-client.crt",
            "severity": "debug"
        },
        "severity": "debug",
        "analyzerSpec": ""
    },
    {
        "name": "host.cerfiticates.collection",
        "labels": {
            "desiredPosition": "1",
            "iconKey": "",
            "iconUri": ""
        },
        "insight": {
            "name": "host.cerfiticates.collection",
            "labels": {
                "iconKey": "",
                "iconUri": ""
            },
            "primary": "Host Cerfiticates Collection",
            "detail": "Certificate is missing, cannot be obtained from /etc/ssl/corp.crt",
            "severity": "error"
        },
        "severity": "error",
        "analyzerSpec": "",
        "error": "Certificate is missing, cannot be obtained from /etc/ssl/corp.crt"
    },
    {
        "name": "host.cerfiticates.collection",
        "labels": {
            "desiredPosition": "1",
            "iconKey": "",
            "iconUri": ""
        },
        "insight": {
            "name": "host.cerfiticates.collection",
            "labels": {
                "iconKey": "",
                "iconUri": ""
            },
            "primary": "Host Cerfiticates Collection",
            "detail": "Certificate is about to expire in 365 days, obtained from /Users/dexteryan/dev/replicated/troubleshoot/ca.crt",
            "severity": "warn"
        },
        "severity": "warn",
        "analyzerSpec": ""
    }
]
```

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [x] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

File: pkg/collect/host_certificate.go [hunk 1]
```diff
@@ -77,3 +116,66 @@ func isEncryptedKey(filename string) (bool, error) {
...
+	}
+
+	return HostCertCollection{
+		CertificatePath:  certPath,
+		CertificateChain: certInfo,
+		Message:          CertValid,
+	}
+}
+
+func HostDecodePem(certInput []byte) tls.Certificate {
```
Comment (banjoh) [inline pkg/collect/host_certificate.go]: No need to duplicate this code. Just use `decodePem`. If you are concerned about `decodePem` function being part of `collect/collect.go`, you can move it to `collect/util.go` like other shared functions.

File: pkg/collect/host_certificate.go [hunk 2]
```diff
@@ -31,43 +55,58 @@ func (c *CollectHostCertificate) IsExcluded() (bool, error) {
...
-			result = KeyPairMismatch
-		} else if strings.Contains(err.Error(), "failed to parse private key") {
-			if encrypted, _ := isEncryptedKey(c.hostCollector.KeyPath); encrypted {
-				result = KeyPairEncrypted
+	if c.hostCollector.CertificatePath != "" {
+		results = append(results, HostCertParser(c.hostCollector.CertificatePath))
+	}
+
+	if c.hostCollector.KeyPath != "" {
+		_, err := tls.LoadX509KeyPair(c.hostCollector.CertificatePath, c.hostCollector.KeyPath)
```
Comment (banjoh) [inline pkg/collect/host_certificate.go]: I don't think we want to care about the private keys. We just want to extract cert meta data

**Review by banjoh — COMMENTED** [review]
I thought I preempt this PR even if its in draft so as to sync a bit on the direction it takes.

I suggest reusing a lot of the certificate metadata extraction code that already exists in the in-cluster certificates collector cause the business logic is the same. I think all other certificate collection types will reuse that code leaving the implementation focus on locating the cerificate/cert-chain.

File: pkg/collect/host_certificate.go (hunk 2 above)
Comment (DexterYan) [reply pkg/collect/host_certificate.go]: Great, I have removed KeyPath, that can simply host collector quite much

File: pkg/collect/host_certificate.go (hunk 1 above)
Comment (DexterYan) [reply pkg/collect/host_certificate.go]: Thank you! I will move it to collect/util.go

File: pkg/apis/troubleshoot/v1beta2/hostcollector_shared.go [hunk 3]
```diff
@@ -165,8 +165,7 @@ type FilesystemPerformance struct {
 
 type Certificate struct {
```
Comment (banjoh) [inline pkg/apis/troubleshoot/v1beta2/hostcollector_shared.go]: I don't think we should break the existing collector. This is a new collector and if we find that we do not need to old one, we should deprecate it

File: pkg/collect/host_certificate.go [hunk 4]
```diff
@@ -2,78 +2,101 @@ package collect
...
+	"encoding/json"
 	"io/ioutil"
 	"path/filepath"
-	"strings"
+	"time"
 
 	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
 )
 
-const KeyPairMissing = "key-pair-missing"
```
Comment (banjoh) [inline pkg/collect/host_certificate.go]: These changes are still needed by the old collector.

**Review by banjoh — COMMENTED** [review]
I did not expect to see any changes in `pkg/collect/host_certificate.go`. This is an existing collector that's used and we do not want to break that functionality. Please create a new file `pkg/collect/host_certificates.go` for any changes related to this host certificates collector.

File: pkg/collect/host_certificate.go (hunk 4 above)
Comment (DexterYan) [reply pkg/collect/host_certificate.go]: Thank you @banjoh, I moved new things into host_certificates.go

**Comment by xavpaice** [comment]
Given that the existing https://troubleshoot.sh/docs/host-collect-analyze/certificate/ `certificate` collector is so close in name, let's rename this one to be a bit more different before we merge it - appreciating that the old one will be deprecated, it'll be a bit of time and it's way too easy to make the mistake of not seeing the 's'.

**Comment by DexterYan** [comment]
I have updated the pr to use certificate validity for host
```
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: certificate
spec:
  hostCollectors:
    - certificateValidity:
        paths:
          - apiserver-kubelet-client.crt
          - /etc/ssl/corp.crt
          - ca.crt
  hostAnalyzers:
    - certificateValidity:
        outcomes:
          - pass:
              message: Certificate is valid
          - warn:
              when: "notAfter < Today + 365 days"
              message: Certificate is about to expire
          - fail:
              when: "notAfter < Today"
              message: Certificate is expired
```
I think it is will be mixed with our existing certificate collector

**Comment by banjoh** [comment]
I would have gone with `certificatesCollection` instead of `certificateValidity`. The nature of the collector allows us to do so much more than just check validity. E.g existence of a root cert, absence of a cert we do not want.

With regards to the name change, it'll now cause a bit of confusion cause now we have `certificates` (in-cluster), `certificate` (host) and `certificateValidity` (host). My opinion here is we'd rather leave with the two that we currently have and accept the risk of confusion, which can can be addressed with [validation of specs](https://github.com/replicatedhq/troubleshoot/issues/871).

**Comment by DexterYan** [comment]
Thank you @banjoh. Agree that certificatesCollection is an overall summary of all certificates, it is more suitable than validity. I have changed in our code

File: pkg/analyze/host_certificates_collection.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	"github.com/replicatedhq/troubleshoot/pkg/collect"
+)
+
+type AnalyzeHostCertificatesCollection struct {
+	hostAnalyzer *troubleshootv1beta2.HostCertificatesCollectionAnalyze
+}
+
+func (a *AnalyzeHostCertificatesCollection) Title() string {
+	return hostAnalyzerTitleOrDefault(a.hostAnalyzer.AnalyzeMeta, "Host Cerfiticates Collection")
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection.go]:
```suggestion
	return hostAnalyzerTitleOrDefault(a.hostAnalyzer.AnalyzeMeta, "Host Certificates Collection")
```

File: pkg/analyze/host_certificates_collection_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+						},
+					},
+				},
+			},
+			result: []*AnalyzeResult{
+				{
+					IsPass:  true,
+					IsWarn:  false,
+					IsFail:  false,
+					Title:   "Host Cerfiticates Collection",
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection_test.go]:
```suggestion
					Title:   "Host Certificates Collection",
```

File: pkg/analyze/host_certificates_collection_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+						},
+					},
+				},
+			},
+			result: []*AnalyzeResult{
+				{
+					IsPass:  false,
+					IsWarn:  false,
+					IsFail:  true,
+					Title:   "Host Cerfiticates Collection",
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection_test.go]:
```suggestion
					Title:   "Host Certificates Collection",
```

File: pkg/analyze/host_certificates_collection_test.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+						},
+					},
+				},
+			},
+			result: []*AnalyzeResult{
+				{
+					IsPass:  false,
+					IsWarn:  true,
+					IsFail:  false,
+					Title:   "Host Cerfiticates Collection",
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection_test.go]:
```suggestion
					Title:   "Host Certificates Collection",
```

File: pkg/analyze/host_certificates_collection_test.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+			}]`,
+			hostAnalyzer: &troubleshootv1beta2.HostCertificatesCollectionAnalyze{
+				Outcomes: []*troubleshootv1beta2.Outcome{},
+			},
+			result: []*AnalyzeResult{
+				{
+					IsPass:  false,
+					IsWarn:  false,
+					IsFail:  true,
+					Title:   "Host Cerfiticates Collection",
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection_test.go]:
```suggestion
					Title:   "Host Certificates Collection",
```

File: pkg/analyze/host_certificates_collection_test.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+						},
+					},
+				},
+			},
+			result: []*AnalyzeResult{
+				{
+					IsPass:  false,
+					IsWarn:  true,
+					IsFail:  false,
+					Title:   "Host Cerfiticates Collection",
```
Comment (CpuID) [inline pkg/analyze/host_certificates_collection_test.go]:
```suggestion
					Title:   "Host Certificates Collection",
```

**Review by CpuID — CHANGES_REQUESTED** [review]
Looks functionally accurate, has test coverage, and the tests are passing.

Few typo suggestions to apply, then I'll approve it 👍

**Review by CpuID — APPROVED** [review]
this is g2g, approved

one minor change in the docs PR https://github.com/replicatedhq/troubleshoot.sh/pull/500 also then I can approve that as well 👍

**Pull request merged**